Core/App.Web/Middleware/ErrorHandlingMiddleware.cs

```diff
@@ -51,32 +62,60 @@ public async Task Invoke(HttpContext context)
         private Task HandleExceptionAsync(HttpContext context, Exception exception)
         {
             var exceptionMessage = GetExceptionMessage(exception);
+            var exceptionType = exception.GetType();
+            var serializedData = JsonSerializer.Serialize(exception.Data);
+            var correlationId = correlationContext.CorrelationContext.CorrelationId;
+            HttpStatusCode exceptionCode = HttpStatusCode.InternalServerError;
             
-            var basicExceptionResult = JsonSerializer.Serialize(new 
-            { 
-                message = exceptionMessage, data = exception.Data 
-            });
-            var fullExceptionResult = JsonSerializer.Serialize(new 
-            { 
-                message = exceptionMessage, data = exception.Data, exception = exception.ToString() 
-            });
 
-            if (_exceptionTypes.TryGetValue(exception.GetType().Name, out HttpStatusCode exceptionCode))
+            foreach (var kv in _exceptionTypes)
+            {
+                if (exceptionType.Equals(kv.Key) || exceptionType.IsSubclassOf(kv.Key))
+                {
+                    exceptionCode = kv.Value;
+                    break;
+                }
+            }
+
+            if (exceptionCode != HttpStatusCode.InternalServerError)
             {
                 // log as warning the expected exceptions
-                _logger.LogWarning($"Expected exception ({(int)exceptionCode})-{exceptionCode}: {basicExceptionResult}"); 
+                _logger.LogWarning(exception, $"Exception: {exceptionType.Name} (expected): HTTP Status: {(int)exceptionCode} ({exceptionCode}): {exceptionMessage} (data: {serializedData})"); 
             }
             else
             {
                 // log as error the unexpected exceptions
                 exceptionCode = HttpStatusCode.InternalServerError; 
-                _logger.LogError(exception, $"Unexpected exception ({(int)exceptionCode})-{exceptionCode}");
+                _logger.LogError(exception, $"Exception: {exceptionType.Name} (unexpected): HTTP Status: {(int)exceptionCode} ({exceptionCode}): {exceptionMessage} (data: {serializedData})");
             }
 
-            string result = _webHostEnvironment.IsDevelopment() ? fullExceptionResult : basicExceptionResult;
             
             context.Response.ContentType = "application/json";
             context.Response.StatusCode = (int)exceptionCode;
+            string result = string.Empty;
+
+            if (_webHostEnvironment.IsDevelopment())
+            {
+                result = JsonSerializer.Serialize(new ErrorResponse
```

Comment:
the HandleException is no longer backwards compatibility. For instance, the data is now missing from the response. I noticed you put in the error details the string output of the GetFullExceptionData(), but I don't understand why do you need to encapsulate everything in a string, instead using the Exception.Data as a Dictionary.
We use a lot Exception.Data on the Allie project. Unfortunately, this change is breaking for our project.

Reply:
@cosmoro do you have an example on how you use it? We can make it backwards compatible if needed.

I like the idea of including the Data as dictionary, however looking for an example on how you use that on the front-end. Also in the past I had sometimes Exceptions which included Data which caused recursion and therefor weren't serializable. Not sure if that still happens though

Reply:
The GetFullExceptionData also deals with InnerException trees. I would love to see an example how you'd like to see a good working result from this.

Reply:
Wrapping in another object shouldn't be a problem, we can adapt our code on FE.
But we always need to have Data as a dictionary on the response, either development or production.
Thanks.

Reply:
The error response should contain:
- message - as retrieved right now (recursively, including the inner exception messages)
- data - as dictionary
- correlationId 
- exception type (could be useful for non-development env)

Additionally the error response for the development env should contain:
- the full exception.
In my opinion, just taking the exception as it was before should be enough.
But if you like to have in a more friendly format you can use the GetFullExceptionData(). As you wish ... but I doubt you display it in your FE app, since it's only used on development.

I consider the additional IsError and Stacktrace fields useless: Stacktrace is already included in the exception, IsError doesn't bring value.

Thank you!

Reply:
@cosmoro I just changed the code on this. I kept the stacktrace and the GetFullExceptionData, this because this will include a tree of inner exceptions when available. I've dealt a number of times with situations where exceptions where returned and the real exception was somewhere in the inner exceptions.

The reason I'm not returning the Exception as well, as I've also dealt with situations where Exceptions weren't serializable (though this is indeed considered bad practice in .NET). However these Exceptions where from Microsoft libraries.

Reply:
Thanks @meghuizen . @cosmoro, I'll mark this as resolved. If needed, we can still adjust before merging these changes into the allie branch.